rspace/src/main/scala/coop/rchain/rspace/ReplayRSpace.scala

```diff
@@ -178,29 +168,32 @@ class ReplayRSpace[F[_]: Concurrent: ContextShift: Log: Metrics: Span, C, P, A,
               comm.consume == source
           }.toSeq),
       c =>
-        store
-          .getData(c)
-          .map(_.iterator.zipWithIndex.toSeq)
-          .map(
-            as => {
-              c -> {
-                if (c == channel)
-                  (Datum(data, persist, produceRef), -1) +: as
-                else as
-              }.filter {
-                matches(comm)
-              }
-            }
-          )
+        for {
+          seqDatum          <- store.getData(c)
+          datumWithIndexSeq = seqDatum.iterator.zipWithIndex.toSeq
+          extendedDatumWithIndexSeq = if (c == channel)
+            (Datum(data, persist, produceRef), -1) +: datumWithIndexSeq
+          else datumWithIndexSeq
+          filteredExtendedDatumWithIndexSeq <- extendedDatumWithIndexSeq.toList.filterA(
+                                                matches(comm)
+                                              )
+        } yield (c, filteredExtendedDatumWithIndexSeq)
     )
 
-  private[this] def matches(comm: COMM)(datumWithIndex: (Datum[A], _)): Boolean = {
+  private[this] def matches(comm: COMM)(datumWithIndex: (Datum[A], _)): F[Boolean] = {
     val datum: Datum[A] = datumWithIndex._1
-    def wasRepeatedEnoughTimes: Boolean =
-      if (!datum.persist) {
-        comm.timesRepeated(datum.source) === produceCounter.get(datum.source)
-      } else true
-    comm.produces.contains(datum.source) && wasRepeatedEnoughTimes
+    def wasRepeatedEnoughTimes: F[Boolean] =
+      for {
+        counter <- produceCounter.get
+      } yield {
+        if (!datum.persist) {
+          comm.timesRepeated(datum.source) === counter(datum.source)
+        } else true
+      }
+
+    for {
+      wasEnough <- wasRepeatedEnoughTimes
+    } yield comm.produces.contains(datum.source) && wasEnough
```

Comment:
Replay tests are now failing because the new code executes `wasEnough` even when the first part of boolean expression if false (produces have datum source).
Hint: use `&&^` function which is lazy in the second argument. This is subtle difference, but it's a good exercise with `F[Boolean]` values.